### ISSUE TITLE
smartcn_tokenizer splits 32-bit Chinese characters into two tokens

### ISSUE THREAD

**Issue by Trey314159** [elastic/elasticsearch#30739]
**Elasticsearch version** 5.5.2

**Plugins installed**: [analysis-hebrew, analysis-icu, analysis-phonetic, analysis-smartcn, analysis-stconvert, analysis-stempel, analysis-ukrainian, experimental-highlighter, extra, extra-analysis-serbian, extra-analysis-slovak, ltr]

**JVM version** (`java -version`): 1.7.0_151

**OS version** (`uname -a` if on a Unix-like system): Linux vagrant 3.16.0-4-amd64 #1 SMP Debian 3.16.43-2 (2017-04-30) x86_64 GNU/Linux

**Description of the problem including expected versus actual behavior**: 32-bit Chinese characters are split into two tokens by `smartcn_tokenizer`. For example, 𨨏 (U+28A0F) is split into two tokens: U+D862 and U+DE0F.

**Steps to reproduce**:

Please include a *minimal* but *complete* recreation of the problem, including
(e.g.) index creation, mappings, settings, query etc.  The easier you make for
us to reproduce it, the more likely that somebody will take the time to look at it.

 1. Create a analyzer "text" as { "type": "custom", "tokenizer": "smartcn_tokenizer"}
 2. `curl -sk localhost:9200/wiki_content/_analyze?pretty -d '{"analyzer": "text", "text" : "𨨏" }'`
 3. Output:
```
{
  "tokens" : [
    {
      "token" : "\uD862",
      "start_offset" : 0,
      "end_offset" : 1,
      "type" : "word",
      "position" : 0
    },
    {
      "token" : "\uDE0F",
      "start_offset" : 1,
      "end_offset" : 2,
      "type" : "word",
      "position" : 1
    }
  ]
}
```
The `standard` tokenizer gives the following, which is the desired token:
```
{
  "tokens" : [
    {
      "token" : "\uD862\uDE0F",
      "start_offset" : 0,
      "end_offset" : 2,
      "type" : "<IDEOGRAPHIC>",
      "position" : 0
    }
  ]
}
```

**Comment by jimczi** [comment]
Thanks for reporting @Trey314159 , I see that somebody (maybe you ?) opened an issue in Lucene with a patch so we'll follow on jira and update the issue here afterward:
https://issues.apache.org/jira/browse/LUCENE-8325

**Comment by Trey314159** [comment]
@jimczi—I wasn't sure whether it was an issue in Elastic or Lucene, so I reported it here. I'm glad someone else narrowed it down to Lucene and maybe found a solution. I've got a workaround hack in Elastic that will work for me for now. Thanks!

**Comment by chengpohi** [comment]
Hi @jimczi & @Trey314159 , I have tried to investigate this issue, and found it's caused by *smartcn* (**lucene**) analyzer, doesn't support the **surrogate** character, when it meets the **surrogate** char it will not tokenize correctly, so for now the solution is correctly handle **surrogate** char boundary to avoid tokenizing to unrecognized char.

**Comment by jimczi** [comment]
Thanks to  @chengpohi the Lucene issue is resolved now and the fix will be available in Elasticsearch 6.4.0 (using Lucene 7.4.0) so I am closing this issue. 
Thanks @Trey314159 for reporting.
